content/2017/03/21/blocking-analytics-spam.rst

```diff
@@ -0,0 +1,145 @@
+months ago we started seeing lots of this:
+
+.. _Google Analytics: https://analytics.google.com/
+
+.. image:: /static/images/blog/2017/ga-spam/ga-spam.jpg
+   :alt: google analytics spam
+   :class: img-border
+
+It's not a new problem, but it's particularly problematic for smaller sites
+that don't receive lots of traffic. On a given day, spam hits were accounting
+for anywhere from ten to ninety (!) percent of our sessions.
+
+.. image:: /static/images/blog/2017/ga-spam/sessions.jpg
+   :alt: google analytics total sessions and filtered (non-spam) sessions
+   :class: img-border
+
+There are `many solutions out there`_; since we mostly saw spam in the
+"referral" field, we wanted a simple way to block spam referrals from being
+included in our analytics data.
+
+.. _many solutions out there: https://www.google.com/#q=how+to+block+google+analytics+spam
+
+
+Blocking Spam Referrals
+-----------------------
+
+One common approach is to blacklist any site visits where ``document.referrer``
+matches a known spam domain. There are `free services`_ that create the
+necessary Google Analytics "filters" for you, but they must be re-configured
+frequently as new spammers are added to the list.
+
+Instead, we tried `spam-referrals-blocker`_, which is a script that blocks
+referrals found on a `community-contributed list of referrer spammers`_. Rather
+than relying on the owner of the script to update it periodically with the
+latest blacklist – or maintaining our own fork of the repo – we decided to
+fetch the latest list as part of our build/deploy process, using `gulp`_ and
+`gulp-download`_:
+
+.. _free services: https://referrerspamblocker.com/
+.. _spam-referrals-blocker: https://github.com/MohamedBassem/spam-referrals-blocker/
+.. _community-contributed list of referrer spammers: https://github.com/piwik/referrer-spam-blacklist
+.. _gulp: http://gulpjs.com/
+.. _gulp-download: https://github.com/Metrime/gulp-download
+
+.. code:: js
+
+    const gulp = require('gulp');
+    const download = require('gulp-download');
+
+    gulp.task('update-spammers', () => {
+      const url = 'https://raw.githubusercontent.com/piwik/referrer-spam-blacklist/master/spammers.txt';
+      return download(url).pipe(gulp.dest('path/to/js/'));
+    });
+
+Once we have an up-to-date blacklist, we import it with the `webpack`_
+`raw-loader`_ and block any referrer found on the list:
+
+.. _webpack: https://webpack.js.org/
+.. _raw-loader: https://github.com/webpack-contrib/raw-loader
+
+.. code:: js
+
+    import spammers from 'raw-loader!./spammers.txt';
+
+    window.isSpamReferral = function () {
+      const list = spammers.split('\n');
+      const currentReferral = document.referrer;
+      if (currentReferral) {
+        for (const spammer of list) {
+          if (spammer && currentReferral.indexOf(spammer) !== -1) {
+            return true;
+          }
+        }
+      }
+      return false;
+    };
```

Comment:
Maybe we should include (or at least mention the potential for) the bit about avoiding analytics events for local domains? It makes the example a bit more complicated, but is a nice bonus feature.